File: src/chipschallenge/blockreactions/ThinWallFrom.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Move.Moves;

public class ThinWallFrom extends BlockReaction {

    private ThinWallFrom() {
    }
    private static ThinWallFrom mInstance = null;

    public static synchronized ThinWallFrom getInstance() {
        if (mInstance == null) {
            mInstance = new ThinWallFrom();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) throws BlockContainerFullException {
        // No reaction;
    }

    public boolean canMove(Block moving, Block standing) {
        switch (moving.getFacing()) {
            case UP:
                return standing.getFacing() != Moves.UP;
            case DOWN:
                return standing.getFacing() != Moves.DOWN;
            case LEFT:
                return standing.getFacing() != Moves.LEFT;
            case RIGHT:
                return standing.getFacing() != Moves.RIGHT;
        }
        return false;
    }
}

File: ChipsChallenge.Shared/Blockreactions/ThinWallFrom.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Moves = Move.Moves;

    public class ThinWallFrom : NoSlipReaction
    {
        private ThinWallFrom()
        {
        }

        private static ThinWallFrom instance;

        public static ThinWallFrom Instance
        {
            get
            {
                lock (typeof(ThinWallFrom))
                {
                    return instance ?? (instance = new ThinWallFrom());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
        }

        public override bool canMove(Block moving, Block standing)
        {
            switch (moving.Facing)
            {
                case Moves.UP:
                    return standing.Facing != Moves.UP;
                case Moves.DOWN:
                    return standing.Facing != Moves.DOWN;
                case Moves.LEFT:
                    return standing.Facing != Moves.LEFT;
                case Moves.RIGHT:
                    return standing.Facing != Moves.RIGHT;
            }
            return false;
        }
    }
}